### PR TITLE
feat: Expansion function

### PR DESCRIPTION
**Pull request by Blakko**
##### Checklist
- [x] `npm test` passes
- [ ] tests and/or benchmarks are included
- [ ] commit message follows commit guidelines

##### Description of change
This adds an `expansion` function to the graph object. This function lets you:
- add multiple items at once
- layout the added items so that they do not overlap (supports animation)

For any improvement, refactoring or change just let me know :)

<details>
  <summary>Examples</summary>
  
https://user-images.githubusercontent.com/2861371/154107301-8134eaf6-3e09-4981-b2a9-43c1c0950603.mp4

https://user-images.githubusercontent.com/2861371/154107314-42edf731-2808-4e01-88bf-964a6ee23b4f.mp4

https://user-images.githubusercontent.com/2861371/154107318-45987927-0a57-4eec-a3e9-9eaca053e34a.mp4

https://user-images.githubusercontent.com/2861371/154107321-0c4892ac-aade-4fa5-a3fc-b86175322883.mp4

https://user-images.githubusercontent.com/2861371/154107322-dbe8493e-5b62-4225-83a3-2d20623b061c.mp4

https://user-images.githubusercontent.com/2861371/154107324-8521c75f-2f12-4e42-a964-9f84a00debe5.mp4
</details>

### CODE REVIEW

**Comment by Yanyan-Wang** [comment]
Maybe add some test cases in the tests/unit ?

File: packages/core/src/graph/graph.ts [hunk 1]
```diff
@@ -1215,6 +1234,295 @@ export default abstract class AbstractGraph extends EventEmitter implements IAbs
...
+        }
+        cumulativeNodesCount = nodesCount;
+      }
+    }
+
+    //
+    // 5. Spread the origin nodes, to make room for the expansion
+    //
+
+    function getPositionAlongTheLine(x1: number, y1: number, x2: number, y2: number, factor: number) {
```
Comment (Yanyan-Wang) [inline packages/core/src/graph/graph.ts:1394]: I think it is something like 'relayout' for force layout, but not suitable for other layout types.

Yes we could assign the initial positions to the newly added nodes in graph.addItems if they have no input positions. But about the relayout part, should be managed by the layout controller and the layout algorithm user configured on the graph

**Comment by Yanyan-Wang** [comment]
So appreciate for your pull request! But I think there are still some problem about function coupling. Please check the comments out and we could have further discussion

File: packages/core/src/graph/graph.ts [hunk 2]
```diff
@@ -1215,6 +1234,295 @@ export default abstract class AbstractGraph extends EventEmitter implements IAbs
...
+          return size / 2 + 108;
+        case 'circle':
+        case 'donut':
+          size = get(item, 'size', 20);
+          return size / 2 + 108;
+        case 'ellipse':
+        case 'rect':
+        case 'modelRect':
+        case 'diamond':
+        case 'image':
```
[the remark this answers is not in the document]
Comment (Blakko) [reply packages/core/src/graph/graph.ts]: I agree this wasn't ideal. I changed the function to use the `INode` bounding box, which makes the firstLayerDistance more accurate and not dependent on the node type.

Let me know if this is ok or if it needs additional refactoring

**Comment by cliffordfajardo** [comment]
@Blakko - is this limited only to node types, does this also support “combo” elements on the graph?

**Comment by Yanyan-Wang** [comment]
Is it deprecated and replaced by https://github.com/antvis/G6/pull/3578 ?

**Comment by Blakko** [comment]
Yes, at this point I'm just using this as a reference

**Pull request closed without merging**